File: typescript-client-sdk/src/sdk/models/operations/getsubscriptionscostsbyid.ts
import { SpeakeasyBase, SpeakeasyMetadata } from "../../../internal/utils";
import * as shared from "../shared";
import { Expose, Transform, Type } from "class-transformer";


export class GetSubscriptionsCostsByIdPathParams extends SpeakeasyBase {
  @SpeakeasyMetadata({ data: "pathParam, style=simple;explode=false;name=subscription_id" })
  subscriptionId: string;
}

export class GetSubscriptionsCostsByIdQueryParams extends SpeakeasyBase {
  @SpeakeasyMetadata({ data: "queryParam, style=form;explode=true;name=group_by" })
  groupBy?: string;

  @SpeakeasyMetadata({ data: "queryParam, style=form;explode=true;name=timeframe_end;dateTimeFormat=YYYY-MM-DDThh:mm:ss.sssZ" })
  timeframeEnd?: Date;

  @SpeakeasyMetadata({ data: "queryParam, style=form;explode=true;name=timeframe_start;dateTimeFormat=YYYY-MM-DDThh:mm:ss.sssZ" })
  timeframeStart?: Date;

  @SpeakeasyMetadata({ data: "queryParam, style=form;explode=true;name=view_mode" })
  viewMode?: string;
}

export class GetSubscriptionsCostsByIdRequest extends SpeakeasyBase {
  @SpeakeasyMetadata()
  pathParams: GetSubscriptionsCostsByIdPathParams;

  @SpeakeasyMetadata()
  queryParams: GetSubscriptionsCostsByIdQueryParams;
}

export class GetSubscriptionsCostsById200ApplicationJSONDataPerPriceCostsPriceGroups extends SpeakeasyBase {
  @SpeakeasyMetadata()
  @Expose({ name: "grouping_key" })
  groupingKey: string;

  @SpeakeasyMetadata()
  @Expose({ name: "grouping_value" })
  groupingValue: string;

  @SpeakeasyMetadata()
  @Expose({ name: "secondary_grouping_key" })
  secondaryGroupingKey?: string;

  @SpeakeasyMetadata()
  @Expose({ name: "secondary_grouping_value" })
  secondaryGroupingValue?: string;

  @SpeakeasyMetadata()
  @Expose({ name: "total" })
  total: string;
}

// GetSubscriptionsCostsById200ApplicationJSONDataPerPriceCosts
/** 
 * Price's contributions for the timeframe, excluding any transforms (minimums and discounts).
**/
export class GetSubscriptionsCostsById200ApplicationJSONDataPerPriceCosts extends SpeakeasyBase {
  @SpeakeasyMetadata()
  @Expose({ name: "price" })
  @Type(() => shared.Price)
  price?: shared.Price;

  @SpeakeasyMetadata({ elemType: GetSubscriptionsCostsById200ApplicationJSONDataPerPriceCostsPriceGroups })
  @Expose({ name: "price_groups" })
  @Type(() => GetSubscriptionsCostsById200ApplicationJSONDataPerPriceCostsPriceGroups)
  priceGroups: GetSubscriptionsCostsById200ApplicationJSONDataPerPriceCostsPriceGroups[];

  @SpeakeasyMetadata()
  @Expose({ name: "subtotal" })
  subtotal?: string;

  @SpeakeasyMetadata()
  @Expose({ name: "total" })
  total?: string;
}

export class GetSubscriptionsCostsById200ApplicationJSONData extends SpeakeasyBase {
  @SpeakeasyMetadata({ elemType: GetSubscriptionsCostsById200ApplicationJSONDataPerPriceCosts })
  @Expose({ name: "per_price_costs" })
  @Type(() => GetSubscriptionsCostsById200ApplicationJSONDataPerPriceCosts)
  perPriceCosts: GetSubscriptionsCostsById200ApplicationJSONDataPerPriceCosts[];

  @SpeakeasyMetadata()
  @Expose({ name: "subtotal" })
  subtotal: string;

  @SpeakeasyMetadata()
  @Expose({ name: "timeframe_end" })
  @Transform(({ value }) => new Date(value), { toClassOnly: true })
  timeframeEnd: Date;

  @SpeakeasyMetadata()
  @Expose({ name: "timeframe_start" })
  @Transform(({ value }) => new Date(value), { toClassOnly: true })
  timeframeStart: Date;

  @SpeakeasyMetadata()
  @Expose({ name: "total" })
  total: string;
}

export class GetSubscriptionsCostsById200ApplicationJSON extends SpeakeasyBase {
  @SpeakeasyMetadata({ elemType: GetSubscriptionsCostsById200ApplicationJSONData })
  @Expose({ name: "data" })
  @Type(() => GetSubscriptionsCostsById200ApplicationJSONData)
  data: GetSubscriptionsCostsById200ApplicationJSONData[];

  @SpeakeasyMetadata()
  @Expose({ name: "pagination_metadata" })
  paginationMetadata: Record<string, any>;
}

export class GetSubscriptionsCostsByIdResponse extends SpeakeasyBase {
  @SpeakeasyMetadata()
  contentType: string;

  @SpeakeasyMetadata()
  statusCode: number;

  @SpeakeasyMetadata()
  getSubscriptionsCostsById200ApplicationJSONObject?: GetSubscriptionsCostsById200ApplicationJSON;
}

File: go-client-sdk/pkg/models/operations/getsubscriptionscostsbyid.go
package operations

import (
	"github.com/speakeasy-sdks/orb-sdks/go-client-sdk/v2/pkg/models/shared"
	"time"
)

type GetSubscriptionsCostsByIDPathParams struct {
	SubscriptionID string `pathParam:"style=simple,explode=false,name=subscription_id"`
}

type GetSubscriptionsCostsByIDQueryParams struct {
	GroupBy        *string    `queryParam:"style=form,explode=true,name=group_by"`
	TimeframeEnd   *time.Time `queryParam:"style=form,explode=true,name=timeframe_end"`
	TimeframeStart *time.Time `queryParam:"style=form,explode=true,name=timeframe_start"`
	ViewMode       *string    `queryParam:"style=form,explode=true,name=view_mode"`
}

type GetSubscriptionsCostsByIDRequest struct {
	PathParams  GetSubscriptionsCostsByIDPathParams
	QueryParams GetSubscriptionsCostsByIDQueryParams
}

type GetSubscriptionsCostsByID200ApplicationJSONDataPerPriceCostsPriceGroups struct {
	GroupingKey            string  `json:"grouping_key"`
	GroupingValue          string  `json:"grouping_value"`
	SecondaryGroupingKey   *string `json:"secondary_grouping_key,omitempty"`
	SecondaryGroupingValue *string `json:"secondary_grouping_value,omitempty"`
	Total                  string  `json:"total"`
}

// GetSubscriptionsCostsByID200ApplicationJSONDataPerPriceCosts
// Price's contributions for the timeframe, excluding any transforms (minimums and discounts).
type GetSubscriptionsCostsByID200ApplicationJSONDataPerPriceCosts struct {
	Price       *shared.Price                                                             `json:"price,omitempty"`
	PriceGroups []GetSubscriptionsCostsByID200ApplicationJSONDataPerPriceCostsPriceGroups `json:"price_groups"`
	Subtotal    *string                                                                   `json:"subtotal,omitempty"`
	Total       *string                                                                   `json:"total,omitempty"`
}

type GetSubscriptionsCostsByID200ApplicationJSONData struct {
	PerPriceCosts  []GetSubscriptionsCostsByID200ApplicationJSONDataPerPriceCosts `json:"per_price_costs"`
	Subtotal       string                                                         `json:"subtotal"`
	TimeframeEnd   time.Time                                                      `json:"timeframe_end"`
	TimeframeStart time.Time                                                      `json:"timeframe_start"`
	Total          string                                                         `json:"total"`
}

type GetSubscriptionsCostsByID200ApplicationJSON struct {
	Data               []GetSubscriptionsCostsByID200ApplicationJSONData `json:"data"`
	PaginationMetadata map[string]interface{}                            `json:"pagination_metadata"`
}

type GetSubscriptionsCostsByIDResponse struct {
	ContentType                                       string
	StatusCode                                        int
	GetSubscriptionsCostsByID200ApplicationJSONObject *GetSubscriptionsCostsByID200ApplicationJSON
}
